src/librustc/hir/mod.rs

```diff
@@ -1059,6 +1071,12 @@ pub enum Expr_ {
     /// For example, `[1; 5]`. The first expression is the element
     /// to be repeated; the second is the number of times to repeat it.
     ExprRepeat(P<Expr>, BodyId),
+
+    /// A suspension point for generators
+    ExprSuspend(P<Expr>),
+
+    /// The argument to a generator
+    ExprImplArg(NodeId),
```

Comment:
Ah, I see what you're hinting at. In my hand-desugared examples I ended up having the generator equivalent of `IntoIterator` produce the first `Yield`/`CoResult`/etc. alongside the generator itself. The other options are multiple entry points, taking `Option<Input>`, or maybe a different suspend model. Maybe a `yield` without a value and produce the first input.

Anyway, what I'm against is any such scheme used with `async` IO, when it can be avoided.